src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,178 @@
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+                int index = (int)(response.Id % (ulong)nodes.Length);
+                UInt160 account = Contract.CreateSignatureRedeemScript(nodes[index]).ToScriptHash();
+                GAS.Mint(engine, account, OracleRequestPrice);
+            }
+        }
+
+        [ContractMethod(OracleRequestPrice, CallFlags.AllowModifyStates)]
+        private void Request(ApplicationEngine engine, string url, string filter, string callback, StackItem userData, long gasForRepsonse)
+        {
+            //Check the arguments
+            if (Utility.StrictUTF8.GetByteCount(url) > MaxUrlLength
+                || (filter != null && Utility.StrictUTF8.GetByteCount(filter) > MaxFilterLength)
+                || Utility.StrictUTF8.GetByteCount(callback) > MaxCallbackLength
+                || gasForRepsonse <= 0)
+                throw new ArgumentException();
+
+            //Mint gas for the response
+            engine.AddGas(gasForRepsonse);
+            GAS.Mint(engine, Hash, gasForRepsonse);
+
+            //Increase the request id
+            StorageItem item_id = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_RequestId));
+            ulong id = BitConverter.ToUInt64(item_id.Value) + 1;
+            item_id.Value = BitConverter.GetBytes(id);
+
+            //Put the request to storage
+            if (engine.Snapshot.Contracts.TryGet(engine.CallingScriptHash) is null)
+                throw new InvalidOperationException();
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Request).Add(item_id.Value), new StorageItem(new OracleRequest
+            {
+                OriginalTxid = GetOriginalTxid(engine),
+                GasForResponse = gasForRepsonse,
+                Url = url,
+                Filter = filter,
+                CallbackContract = engine.CallingScriptHash,
+                CallbackMethod = callback,
+                UserData = BinarySerializer.Serialize(userData, MaxUserDataLength)
+            }));
+
+            //Add the id to the IdList
+            engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)), () => new StorageItem(new IdList())).GetInteroperable<IdList>().Add(id);
```

Comment:
we share the url by different request, but also by different responses, it could collide with a different response with the same url, but different content.

Reply:
If the same URL returns different content, then this URL is not suitable for Oracle. But this is not a big problem, because its behavior in the contract is consistent.

Reply:
> If the same URL returns different content, then this URL is not suitable for Oracle

The price of btc for example, it could return different values. https://api.coinpaprika.com/v1/coins/btc-bitcoin

Reply:
A URL suitable for Oracle should be time-dependent or with ID, such as:

```
https://api.xxxxx.com/v1/coins/btc-bitcoin?t=1595064604
https://api.xxxxx.com/v1/coins/btc-bitcoin?id=100
```

It should return the same content if the timestamps or the IDs in the URl are the same.

Reply:
I see, @erikzhang, and I also understand what @shargon wants to say.

In fact, a time-dependent URL can be an option. On the other hand, in a first moment, I imagined that Oracles would return that contents and smart contracts would handle responses even if they are distinct.

In fact, since we are not going to create a generic template for oracles, as we discussed before, I believe that the fixed time-dependent may be enough. Last time we discussed Erik said that other oracle services should register their own Smart Contract.

In a first view of the Oracle Service I imagined that we would provide a synchronous service that any third party would be able to register and offer they services integrated with Neo Blockchain by design.